src/promptflow/promptflow/_sdk/_record_storage.py

```diff
@@ -0,0 +1,228 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import collections
+import hashlib
+import json
+from pathlib import Path
+from typing import Any, Dict, OrderedDict
+
+from promptflow._sdk._configuration import Configuration
+from promptflow._sdk._constants import RecordMode
+from promptflow._sdk._errors import RecordFileMissingException, RecordItemMissingException
+from promptflow._utils.dataclass_serializer import serialize
+from promptflow._utils.tool_utils import get_inputs_for_prompt_template
+
+
+class RecordStorage(object):
+    """
+    RecordStorage is used to store the record of node run.
+    File often stored in .promptflow/node_cache.json
```

Comment:
document why we chose this implementation. persist the data as json by ourself. 
Do we have any forsee limitations here?

Reply:
No if there is always text based llm records.
Yes if there is more llm types like mp3/image.

We could extend the Record storage with further implementations.